### PR TITLE
Do not copy messages in memory before saving them to the blob store

### PR DESCRIPTION
**Pull request by chibenwa**
- This reduce memory impact for enqueueing emails - SMTP, JMAP message submission
 - This reduce memory impact of dequeuing emails - mailetcontainer processing
 - This reduce overhead of CassandraMailRepository (minor gains)

 The memory gain will be major upon big messages.

 Inefficient memory usage of MailboxMessage is not addressed here.

Here is a short presentation of the pattern and concept underneath that I did for my team:

[efficient-io.pdf](https://github.com/apache/james-project/files/6073927/efficient-io.pdf)

### CODE REVIEW

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
The message is then represented as a byte array only when reading it from the mailbox...

I am running some SMTP benchmarks on this...

**Comment by chibenwa** [comment]
1st commit is :green_apple:  on the CI

I run performance tests for it...

## Before

Capture d’écran de 2021-02-26 08-58-01

Capture d’écran de 2021-02-26 08-57-58


## After

Capture d’écran de 2021-02-26 08-58-07

Capture d’écran de 2021-02-26 08-58-04

## Conclusion

Little performance enhancement. Great p99 diminition. More importantly we can notice that no "stop the world GC was triggered" during long period of time, as in the first run.

I will continue looking for places where to avoid memory copies, will fix compilation of the last commit, and continue benches ;-)

**Comment by chibenwa** [comment]
As off 6743c2a 

Capture d’écran de 2021-02-26 23-26-47

**Comment by chibenwa** [comment]
Broken tests:

```
org.apache.james.transport.mailets.AddFooterTest.shouldAddFooterWhenMultipartMessage
    org.apache.james.transport.mailets.ICSAttachmentWorkflowTest.mailShouldNotContainCalendarContentInTextBodyButAttachment

& JPA MPT tests
```

I will investigate them...

Perf wise:

Capture d’écran de 2021-02-28 06-22-42

**Comment by chibenwa** [comment]
Here we are! Finnally some real improvments!

Capture d’écran de 2021-02-28 13-51-37

Almost x3 throughtput on SMTP tests!

Interestingly, running `$ dstat -lrvn 10` on the testing server:

```
---load-avg--- --io/total- ---procs--- ------memory-usage----- ---paging-- -dsk/total- ---system-- ----total-cpu-usage---- -net/total-
 1m   5m  15m | read  writ|run blk new| used  buff  cach  free|  in   out | read  writ| int   csw |usr sys idl wai hiq siq| recv  send
10.5 10.6 10.5|47.2   339 |6.5 0.5 2.6|15.4G 4240k 13.1G  262M|   0     0 | 760k   48M|  21k   35k| 31  12  53   3   0   2|  25M   38M
9.28 10.3 10.4|39.1   474 |3.2 0.5 2.9|15.5G 4144k 13.1G  228M|   0     0 | 593k   49M|  16k   26k| 24  10  62   3   0   1|  22M   32M
10.3 10.5 10.5|48.3   617 |6.1 0.1  12|15.5G 3884k 13.1G  249M|   0     0 | 743k   62M|  22k   38k| 31  14  52   2   0   2|  28M   41M
9.30 10.3 10.4|48.1   318 |4.8 0.5 4.7|15.5G 3808k 13.1G  231M|   0     0 | 914k   48M|  18k   30k| 28  11  57   2   0   1|  23M   34M
8.86 10.1 10.3|43.2   315 |5.6 0.2 2.9|15.4G 3716k 13.0G  323M|   0     0 | 732k   47M|  19k   33k| 30  12  55   2   0   2|  25M   37M
```

We can notice that we are massively writing to the disk (~50MB/s) but almost not reading. Previous runs were hitting 50MB/s reads + ~100MB/s writes. I assume we now better leverage the FileSystem cache.

**Comment by chibenwa** [comment]
```
Test Result (5 failures / -74)
    org.apache.james.mailrepository.jdbc.JDBCMailRepositoryTest.storingMessageWithPerRecipientHeadersShouldAllowMultipleHeadersPerUser
    org.apache.james.mailrepository.jdbc.JDBCMailRepositoryTest.retrieveShouldGetStoredMail
    org.apache.james.mailrepository.jdbc.JDBCMailRepositoryTest.storingMessageWithSameKeyTwiceShouldUpdateMessageContent
    org.apache.james.mailrepository.jdbc.JDBCMailRepositoryTest.storingMessageWithSameKeyTwiceShouldUpdateMessageAttributes
    org.apache.james.mailrepository.jdbc.JDBCMailRepositoryTest.retrieveShouldReturnAllMailProperties
```

The freaks are out ;-)

I also have some JMAP benchs:

Before:

Capture d’écran de 2021-02-28 18-49-35

After:

Capture d’écran de 2021-02-28 18-49-38

Sending emails via JMAP is thus significantly faster (better p99, meantime and throughtput)

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java [hunk 1]
```diff
@@ -171,6 +172,16 @@ public void shouldCacheWhenEmptyStream() {
         });
     }
 
+    @Test
+    public void shouldCacheWhenEmptyByteSource() {
+        BlobId blobId = Mono.from(testee().save(DEFAULT_BUCKETNAME, ByteSource.wrap(EMPTY_BYTEARRAY), SIZE_BASED)).block();
+
+        SoftAssertions.assertSoftly(soflty -> {
```
Comment (Arsnael) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java:179]: s/soflty/softly

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java [hunk 2]
```diff
@@ -193,6 +204,18 @@ public void shouldCacheWhenFiveKilobytesSteam() {
         });
     }
 
+    @Test
+    public void shouldCacheWhenFiveKilobytesByteSource() {
+        BlobId blobId = Mono.from(testee().save(DEFAULT_BUCKETNAME, ByteSource.wrap(APPROXIMATELY_FIVE_KILOBYTES), SIZE_BASED)).block();
+
+        SoftAssertions.assertSoftly(soflty -> {
```
Comment (Arsnael) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java:211]: idem

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java [hunk 3]
```diff
@@ -292,6 +315,30 @@ public void shouldNotCacheWhenReadWithBigStream() {
         });
     }
 
+    @Test
+    public void shouldNotCacheWhenReadWithOutDefaultBucketByteSource() {
+        BlobId blobId = Mono.from(backend.save(TEST_BUCKETNAME, ByteSource.wrap(APPROXIMATELY_FIVE_KILOBYTES), SIZE_BASED)).block();
+
+        SoftAssertions.assertSoftly(soflty -> {
```
Comment (Arsnael) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java:322]: idem

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java [hunk 4]
```diff
@@ -292,6 +315,30 @@ public void shouldNotCacheWhenReadWithBigStream() {
...
+                .hasSameContentAs(new ByteArrayInputStream(APPROXIMATELY_FIVE_KILOBYTES));
+            soflty.assertThat(Mono.from(cache.read(blobId)).blockOptional()).isEmpty();
+        });
+    }
+
+    @Test
+    public void shouldNotCacheWhenReadWithBigByteSource() {
+        BlobId blobId = Mono.from(backend.save(DEFAULT_BUCKETNAME, ByteSource.wrap(TWELVE_MEGABYTES), SIZE_BASED)).block();
+
+        SoftAssertions.assertSoftly(soflty -> {
```
Comment (Arsnael) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java:334]: idem

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageWrapper.java [hunk 5]
```diff
@@ -361,13 +361,8 @@ public int getSize() throws MessagingException {
                 throw new MessagingException("Unable to calculate message size");
             }
         } else {
-            if (!messageParsed) {
-                loadMessage();
-            }
-
-            return super.getSize();
+            return -1;
```
Comment (Arsnael) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageWrapper.java:364]: -1 standing for what here?

File: server/container/core/src/test/java/org/apache/james/server/core/MimeMessageWrapperTest.java [hunk 6]
```diff
@@ -256,15 +256,15 @@ public void testSizeModifiedHeaders() throws MessagingException {
...
+        assertThat(mw.getSize()).isEqualTo(-1);
     }
 
     @Test
     public void testSizeModifiedBodyWithSave() throws MessagingException {
         String newBody = "This is the new body of the message";
         mw.setText(newBody);
         mw.saveChanges();
-        assertThat(mw.getSize()).isEqualTo(body.length());
+        assertThat(mw.getSize()).isEqualTo(-1);
```
Comment (Arsnael) [inline server/container/core/src/test/java/org/apache/james/server/core/MimeMessageWrapperTest.java:267]: Are those test names still making sense with that -1 returned?

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageWrapper.java (hunk 5 above)
Comment (chibenwa) [reply server/container/core/src/main/java/org/apache/james/server/core/MimeMessageWrapper.java:364]: UNKNOWN

File: server/container/core/src/test/java/org/apache/james/server/core/MimeMessageWrapperTest.java (hunk 6 above)
Comment (chibenwa) [reply server/container/core/src/test/java/org/apache/james/server/core/MimeMessageWrapperTest.java:267]: I think as they don't tell anything about what behaviour is expected but just what part of what is tested...

**Comment by jeantil** [comment]
@chibenwa it's quite big, I couldn't complete the review yet but I wanted to let you know I'm highly interested.

**Comment by chibenwa** [comment]
> @chibenwa it's quite big, I couldn't complete the review yet but I wanted to let you know I'm highly interested.

Quite big but I did try to keep a clear history...

**Comment by chibenwa** [comment]
Adding  https://github.com/apache/james-project/pull/211 to the mix:

Screenshot from 2021-03-01 17-01-40

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobStore.java [hunk 7]
```diff
@@ -35,6 +37,8 @@
 
     Publisher<BlobId> save(BucketName bucketName, InputStream data, StoragePolicy storagePolicy);
 
+    Publisher<BlobId> save(BucketName bucketName, ByteSource data, StoragePolicy storagePolicy);
```
Comment (chibenwa) [inline server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobStore.java:40]: Maybe we need an API equivalent to `ByteSource` of our own here (called eg `Blob`) providing content and length.

Feedback?

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java [hunk 8]
```diff
@@ -210,6 +212,23 @@ public InputStream read(BucketName bucketName, BlobId blobId) {
         return backend.save(bucketName, inputStream, storagePolicy);
     }
 
+    @Override
+    public Publisher<BlobId> save(BucketName bucketName, ByteSource byteSource, StoragePolicy storagePolicy) {
+        Preconditions.checkNotNull(byteSource, "ByteSource must not be null");
+
+        if (isAbleToCache(bucketName, storagePolicy)) {
+            return Mono.from(backend.save(bucketName, byteSource, storagePolicy))
```
Comment (mbaechler) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java:220]: why don't we call that `saveInCache` as for other `save` methods?

File: server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala [hunk 9]
```diff
@@ -51,6 +52,15 @@ class PassThroughBlobStore @Inject()(blobStoreDAO: BlobStoreDAO,
       .`then`(SMono.just(blobId))
   }
 
+  override def save(bucketName: BucketName, data: ByteSource, storagePolicy: BlobStore.StoragePolicy): Publisher[BlobId] = {
```
Comment (mbaechler) [inline server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala:55]: usually, starting at the third copy/paste, I factorize the code.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AttachmentMapperTest.java [hunk 10]
```diff
@@ -159,7 +160,7 @@ void getRelatedMessageIdsShouldReturnMessageIdWhenStoredWithMessageId() throws E
         MessageId messageId = generateMessageId();
         AttachmentId attachmentId = attachmentMapper.storeAttachmentsForMessage(ImmutableList.of(ParsedAttachment.builder()
             .contentType("content")
-            .content("".getBytes(StandardCharsets.UTF_8))
+            .content(ByteSource.wrap("".getBytes(StandardCharsets.UTF_8)))
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AttachmentMapperTest.java:163]: it's a very complex way of providing an empty ByteSource

File: server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java [hunk 11]
```diff
@@ -135,6 +134,18 @@ public Response onLine(SMTPSession session, ByteBuffer lineByteBuffer, LineHandl
         return null;
     }
 
+    private byte[] readBytes(ByteBuffer line) {
```
Comment (mbaechler) [inline server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java:137]: there's no existing method to read a bytebuffer to an array?

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java [hunk 12]
```diff
@@ -121,6 +121,9 @@ private ByteBuffer toByteBuffer(byte[] bytes) {
 
     private byte[] toByteArray(Row row) {
         ByteBuffer byteBuffer = row.getBytes(DATA);
+        if (byteBuffer.hasArray()) {
+            return byteBuffer.array();
```
Comment (mbaechler) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:125]: there's no guarantee that:
* the buffer won't be recycled by another layer
* the array has the size of the data

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/delivery/MailboxAppender.java [hunk 13]
```diff
@@ -78,8 +77,7 @@ private AppendResult append(MimeMessage mail, Username user, String folder, Mail
     }
 
     private AppendResult appendMessageToMailbox(MimeMessage mail, MailboxSession session, MailboxPath path) throws MailboxException, MessagingException {
-        createMailboxIfNotExist(session, path);
-        final MessageManager mailbox = mailboxManager.getMailbox(path, session);
+        MessageManager mailbox = createMailboxIfNotExist(session, path);
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/delivery/MailboxAppender.java:85]: can we put that commit into another PR please?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 14]
```diff
@@ -321,27 +321,17 @@ public AppendResult appendMessage(InputStream msgIn, Date internalDate, final Ma
             // source for the InputStream
             file = File.createTempFile("imap", ".msg");
             try (FileOutputStream out = new FileOutputStream(file);
-                 BufferedOutputStream bufferedOut = new BufferedOutputStream(out);
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java:324]: can we put this commit in another PR?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java [hunk 15]
```diff
@@ -242,6 +242,11 @@ public Builder withInternalDate(Date date) {
                 return this;
             }
 
+            public Builder withInternalDate(Optional<Date> date) {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java:252]: can we put this commit in another PR please?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 16]
```diff
@@ -439,9 +470,10 @@ private int getBodyStartOctet(BodyOffsetInputStream bIn) {
         return bodyStartOctet;
     }
 
-    private AppendResult createAndDispatchMessage(Date internalDate, MailboxSession mailboxSession, File file, PropertyBuilder propertyBuilder, Flags flags, int bodyStartOctet) throws IOException, MailboxException {
-        try (SharedFileInputStream contentIn = new SharedFileInputStream(file)) {
-            final int size = (int) file.length();
+    private AppendResult createAndDispatchMessage(Date internalDate, MailboxSession mailboxSession, Content content, PropertyBuilder propertyBuilder, Flags flags, int bodyStartOctet) throws IOException, MailboxException {
+        // TODO MailboxMessage should be backed by a "content"
+        try (SharedFileInputStream contentIn = null) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: this commit breaks that code on purpose? is it fixed later?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 17]
```diff
@@ -347,6 +360,24 @@ public AppendResult appendMessage(InputStream msgIn, Date internalDate, final Ma
         }
     }
 
+    public AppendResult appendMessage(Content msgIn, Date internalDate, final MailboxSession mailboxSession, boolean isRecent, Flags flagsToBeSet) throws MailboxException {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java:362]: that's quite a bunch of logic, is it tested in some way?

File: server/protocols/protocols-lmtp/src/main/java/org/apache/james/lmtpserver/hook/MailboxDeliverToRecipientHandler.java [hunk 18]
```diff
@@ -80,15 +82,25 @@ public HookResult deliver(SMTPSession session, MailAddress recipient, MailEnvelo
             mailboxManager.getMailbox(MailboxPath.inbox(username), mailboxSession)
                 .appendMessage(MessageManager.AppendCommand.builder()
                     .recent()
-                    .build(envelope.getMessageInputStream()),
+                    .build(new Content() {
```
Comment (mbaechler) [inline server/protocols/protocols-lmtp/src/main/java/org/apache/james/lmtpserver/hook/MailboxDeliverToRecipientHandler.java:85]: I didn't go back to check all `Content` instances but it looks to be a lot of inline classes definitions that are exactly the same

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java [hunk 19]
```diff
@@ -116,90 +116,90 @@ public void shouldCacheWhenDefaultBucketName() {
     public void shouldNotCacheWhenNotDefaultBucketName() {
         BlobId blobId = Mono.from(testee().save(TEST_BUCKETNAME, EIGHT_KILOBYTES, SIZE_BASED)).block();
 
-        SoftAssertions.assertSoftly(soflty -> {
-            soflty.assertThat(Mono.from(cache.read(blobId)).blockOptional()).isEmpty();
-            soflty.assertThat(Mono.from(backend.readBytes(TEST_BUCKETNAME, blobId)).block()).containsExactly(EIGHT_KILOBYTES);
+        SoftAssertions.assertSoftly(softly -> {
```
Comment (mbaechler) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/cache/CachedBlobStoreTest.java:119]: could you move this commit to another PR please?

File: server/container/core/src/main/java/org/apache/james/server/core/MimeMessageWrapper.java [hunk 20]
```diff
@@ -417,7 +417,7 @@ public long getMessageSize() throws MessagingException {
             }
         } else if (source != null && !bodyModified) {
             try  (InternetHeadersInputStream headerStream = new InternetHeadersInputStream(headers)) {
-                return source.getMessageSize() - initialHeaderSize + IOUtils.consume(headerStream);
+                return source.getMessageSize() - initialHeaderSize + IOUtils.consume(headerStream) -2;
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/server/core/MimeMessageWrapper.java:420]: please avoid such magic number

File: server/container/core/src/test/java/org/apache/james/server/core/MimeMessageWrapperTest.java [hunk 21]
```diff
@@ -292,4 +294,14 @@ public void saveChangesShouldPreserveMessageId() throws Exception {
         assertThat(mimeMessageWrapper.getMessageID())
             .isEqualTo(messageId);
     }
+
+    @Test
+    public void test() throws Exception {
```
Comment (mbaechler) [inline server/container/core/src/test/java/org/apache/james/server/core/MimeMessageWrapperTest.java:299]: could we have a proper test name?

**Review by mbaechler — COMMENTED** [review]
This PR is really interesting.

However, I would like all non-directly related changes to be moved into a PR to make this one more accessible to people wanting to read it.

Do you think it's manageable?
Also, it may also help to squash fixups.

**Comment by chibenwa** [comment]
>However, I would like all non-directly related changes to be moved into a PR to make this one more accessible to people wanting to read it.

Most of the changes are related.

File: server/container/core/src/test/java/org/apache/james/server/core/MimeMessageWrapperTest.java (hunk 21 above)
Comment (chibenwa) [reply server/container/core/src/test/java/org/apache/james/server/core/MimeMessageWrapperTest.java:299]: Yes for sure.

File: server/protocols/protocols-lmtp/src/main/java/org/apache/james/lmtpserver/hook/MailboxDeliverToRecipientHandler.java (hunk 18 above)
Comment (chibenwa) [reply server/protocols/protocols-lmtp/src/main/java/org/apache/james/lmtpserver/hook/MailboxDeliverToRecipientHandler.java:85]: a lot of inline classes definitions that looks exactly the same. Nuance. I generalized where possible.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 17 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java:362]: Yes in Manager tests, refactorings were automated.

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java (hunk 12 above)
Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:125]: We likely can drop this commit if you wish.

File: server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java (hunk 11 above)
Comment (chibenwa) [reply server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java:137]: Nope not that I know>

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AttachmentMapperTest.java (hunk 10 above)
Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AttachmentMapperTest.java:163]: But it works and is simple to understand / write.

You have an alternative?

File: server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala (hunk 9 above)
Comment (chibenwa) [reply server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala:55]: I'd say let's drop the InputSream method.

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java (hunk 8 above)
Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java:220]: Because the params are not exactly the same.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 16 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Yes. And it get fixed later. (I assume it).

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java (hunk 15 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java:252]: No it is directly related.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 14 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java:324]: No it is directly related to me.

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java (hunk 12 above)
Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:125]: Dropped

Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java]: BTW you are wrong:

https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#hasArray()

```
Tells whether or not this buffer is backed by an accessible byte array.

If this method returns true then the array and arrayOffset methods may safely be invoked. 
```

https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#array()

```
Returns the byte array that backs this buffer  (optional operation). 
```

As I understand that you feel concerned I will open a separated PR, But still want this code merged.

Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:125]: see https://github.com/apache/james-project/pull/316

File: server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala (hunk 9 above)
Comment (chibenwa) [reply server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala:55]: Fixups welcomed ;-)

The code is simple, easy to understand, tested, vary subtly in inputs and cannot easily find a way to factorize it hence will not spend my time on it.

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java (hunk 8 above)
Comment (mbaechler) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java:220]: I mean, you could create a method named `saveInCache` to make the code clearer _as we did for other methods_

File: server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala (hunk 9 above)
Comment (mbaechler) [reply server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala:55]: > Fixups welcomed ;-)

Come on, if you don't want the code reviewed, just say it.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AttachmentMapperTest.java (hunk 10 above)
Comment (mbaechler) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AttachmentMapperTest.java:163]: `ByteSource.empty()` ?

File: server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java (hunk 11 above)
Comment (mbaechler) [reply server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java:137]: me neither

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java (hunk 12 above)
Comment (mbaechler) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:125]: Let's say I'm wrong (because you look really sure of your opinion). Then open ByteBuffer.wrap and tell me what happen when offset/limit is used. Do you see any shrinking of the buffer somewhere?

I can remember being bitten by this in the past, please look closer to this problem because I'm confident it's not correct code.

Comment (chibenwa) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CassandraBlobStoreCache.java:125]: To be fair, from what I see we could use the offset and the length to generate a stream on demand without an array copy.

Here the callers need a byte[] access.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AttachmentMapperTest.java (hunk 10 above)
Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AttachmentMapperTest.java:163]: Fair

File: server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala (hunk 9 above)
Comment (chibenwa) [reply server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/PassThroughBlobStore.scala:55]: Come on, I ask kindly for suggestions, and you give none...

**Comment by chibenwa** [comment]
:green_apple: build

I will rebase, clean up the history and undraft this

**Comment by chibenwa** [comment]
:green_apple: build.

I would like to merge this reasonably soon as is ships nice enhancements.

**Comment by Arsnael** [comment]
> I would like to merge this reasonably soon as is ships nice enhancements.

Then undraft it please. As long as it's in draft state it's not clear if the work is finished or not

File: server/data/data-jdbc/src/test/java/org/apache/james/mailrepository/jdbc/JDBCMailRepositoryTest.java [hunk 22]
```diff
@@ -79,21 +79,21 @@ private BasicDataSource getDataSource() {
 
     @Override
     @Disabled("JAMES-2546 This mail repository does not support null sender")
-    public void storeRegularMailShouldNotFailWhenNullSender() {
-
+    public void storeRegularMailShouldNotFailWhenNullSender() throws Exception {
+        MailRepositoryContract.super.storeRegularMailShouldNotFailWhenNullSender();
```
Comment (Arsnael) [inline server/data/data-jdbc/src/test/java/org/apache/james/mailrepository/jdbc/JDBCMailRepositoryTest.java:83]: If it's disabled it does not matter to have a body, as it will not be played. My previous comment was for the one with Unstable Tag, as those tests are still played, but in a separate phase of the build

Comment (chibenwa) [reply server/data/data-jdbc/src/test/java/org/apache/james/mailrepository/jdbc/JDBCMailRepositoryTest.java:83]: It does not hurt to have a body when disabled as far as I know

Comment (Arsnael) [reply server/data/data-jdbc/src/test/java/org/apache/james/mailrepository/jdbc/JDBCMailRepositoryTest.java:83]: It does not, just useless. But I don't mind TBH

**Review by LanKhuat — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged.

If there is some refactorings to do on this work I can handle them in another PR.
